autora/controller/state/history.py

```diff
@@ -0,0 +1,613 @@
+            >>> h.conditions
+            [(1, 2, 3), (4, 5, 6), (7, 8, 9)]
+
+        """
+        return self._by_kind.conditions
+
+    @property
+    def observations(self) -> List[ArrayLike]:
+        """
+
+        Returns:
+
+        Examples:
+            The sequence of all observations is returned
+            >>> h = History(observations=["1st observation"])
+            >>> h.observations
+            ['1st observation']
+
+            >>> h = h.update(observations=["2nd observation"])
+            >>> h.observations  # doctest: +ELLIPSIS
+            ['1st observation', '2nd observation']
+
+        """
+        return self._by_kind.observations
+
+    @property
+    def theories(self) -> List[BaseEstimator]:
+        """
+
+        Returns:
+
+        Examples:
+            View the sequence of theories with one theory:
+            >>> s = History(theories=["1st theory"])
+            >>> s.theories  # doctest: +NORMALIZE_WHITESPACE
+            ['1st theory']
+
+            ... or more theories:
+            >>> s = s.update(theories=["2nd theory"])  # doctest: +NORMALIZE_WHITESPACE
+            >>> s.theories
+            ['1st theory', '2nd theory']
+
+        """
+        return self._by_kind.theories
+
+    @property
+    def history(self) -> List[Result]:
+        """
+
+        Examples:
+            We initialze some history:
+            >>> h = History(theories=['t1', 't2'], conditions=['c1', 'c2'],
+            ...     observations=['o1', 'o2'], params={'a': 'param'}, metadata=VariableCollection(),
+            ...     history=[Result("from history", ResultKind.METADATA)])
+
+            Parameters passed to the constructor are included in the history in the following order:
+            `history`, `metadata`, `params`, `conditions`, `observations`, `theories`
+
+            >>> h.history  # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+            [Result(data='from history', kind=ResultKind.METADATA),
+             Result(data=VariableCollection(...), kind=ResultKind.METADATA),
+             Result(data={'a': 'param'}, kind=ResultKind.PARAMS),
+             Result(data='c1', kind=ResultKind.CONDITION),
+             Result(data='c2', kind=ResultKind.CONDITION),
+             Result(data='o1', kind=ResultKind.OBSERVATION),
+             Result(data='o2', kind=ResultKind.OBSERVATION),
+             Result(data='t1', kind=ResultKind.THEORY),
+             Result(data='t2', kind=ResultKind.THEORY)]
+
+            If we add a new value, like the params object, the updated value is added to the
+            end of the history:
+            >>> h = h.update(params={'new': 'param'})
+            >>> h.history  # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+            [..., Result(data={'new': 'param'}, kind=ResultKind.PARAMS)]
+
+        """
+        return self._history
+
+    def filter_by(self, kind: Optional[Set[Union[str, ResultKind]]] = None) -> History:
```

Comment:
Might be nice to also be able to filter by tag. For instance if you wanted to find the conditions, observations, and theory of a certain cycle. For that case would you need to label the data tag in a sensible way?

Reply:
On reflection, I think we shouldn't do this. Instead, we'll say that people should subclass the `Result` object (or roll their own, it's a super simple object) if they want to add more metadata fields.